### PR TITLE
Strip common subdomains when matching incoming toolbar queries.

### PR DESCRIPTION
**Pull request by MatthewTighe**
Closes [Fenix #22555](https://github.com/mozilla-mobile/fenix/issues/22555). Should this change be included in the changelog, or is the log geared more for API changes that could affect downstream? Not seeing too many bug fixes in there

Also curious as to whether this might be a change that should be propagated to A-S. We seem to handle a few edge cases that aren't covered by the tests that I saw in A-S, but they all seem like potentially useful behavior to share. This issue was discovered as a behavior difference between Android and desktop, for example.

### Pull Request checklist
- [x] **Quality**: This PR builds and passes detekt/ktlint checks
- [x] **Tests**: This PR includes tests
- [ ] **Changelog**: See above question
- [x] **Accessibility**: No significant user-facing feature changes

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: buildSrc/src/main/java/Dependencies.kt [hunk 1]
```diff
@@ -54,7 +54,7 @@ object Versions {
         const val test = "1.3.0"
         const val test_ext = "1.1.2"
         const val espresso = "3.3.0"
-        const val room = "2.3.0"
+        const val room = "2.4.0-beta02"
```
Comment (Amejia481) [inline buildSrc/src/main/java/Dependencies.kt:57]: Just to confirm that this changes was intentional ? :)

File: components/support/utils/src/test/java/mozilla/components/support/utils/DomainMatcherTest.kt [hunk 2]
```diff
@@ -61,6 +67,11 @@ class DomainMatcherTest {
             DomainMatch("http://firefox.com", "firefox.com"),
             segmentAwareDomainMatch("fire", urls)
         )
+//        // Path segments
+//        assertEquals(
+//            DomainMatch("bcferries.com/current-conditions", "bcferries.com/current-conditions"),
+//            segmentAwareDomainMatch("bcferries.com/cu", urls)
+//        )
```
Comment (Amejia481) [inline components/support/utils/src/test/java/mozilla/components/support/utils/DomainMatcherTest.kt:74]: Just to confirm that this change was intentional ? :)

File: buildSrc/src/main/java/Dependencies.kt (hunk 1 above)
Comment (MatthewTighe) [reply buildSrc/src/main/java/Dependencies.kt:57]: Ah, my bad. This is what I get for submitted a PR right at the end of my day.

This change doesn't need to be accepted, but is required for A-C to build correctly on M1 macs. What level of care is generally taken for updating dependencies in this project? Are we more careful about alpha and beta artifacts?

File: components/support/utils/src/test/java/mozilla/components/support/utils/DomainMatcherTest.kt (hunk 2 above)
Comment (MatthewTighe) [reply components/support/utils/src/test/java/mozilla/components/support/utils/DomainMatcherTest.kt:74]: This one was very unintentional 😂 Thanks for calling it out!

File: buildSrc/src/main/java/Dependencies.kt (hunk 1 above)
Comment (pocmo) [reply buildSrc/src/main/java/Dependencies.kt:57]: Yeah, we try to only use stable dependencies. Other than that, updating to the latest stable version is great - but needs to be thoroughly tested since it will update it for downstream projects too.

Comment (MatthewTighe) [reply buildSrc/src/main/java/Dependencies.kt:57]: Cool, that makes sense. I'll revert this change

**Comment by MatthewTighe** [comment]
After spending a bit too long hacking at this trying to get the tests you wrote to pass, I finally realized that we actually don't need to handle path segment matching in `DomainMatcher`. The matching behavior provided by A-S actually already does it. 

That is, by the time a URL is received by `DomainMatcher`, it's already the closest subset of path. You can see that demonstrated in the tests you and I both wrote in `PlacesHistoryStorageTest` (which already pass):

```
        res = history.getAutocompleteSuggestion("en.wikipedia.org/wi")!!
        assertEquals("en.wikipedia.org/wiki/", res.text)
        assertEquals("https://en.wikipedia.org/wiki/", res.url)
        assertEquals("placesHistory", res.source)
        assertEquals(1, res.totalItems)

        history.recordVisit("https://www.github.com/mozilla-mobile/fenix", PageVisit(VisitType.LINK))
        res = history.getAutocompleteSuggestion("github.com/moz")!!
        assertEquals("github.com/mozilla-mobile/", res.text)
        assertEquals("https://www.github.com/mozilla-mobile/", res.url)
        assertEquals("placesHistory", res.source)
        assertEquals(1, res.totalItems)
```

It seems like the functionality of `DomainMatcher` is really just to strip extraneous data from a fully normalized URL like the protocol and subdomains. We could still replicate the path segment matching behavior there as well, but it seems redundant to me.

That said, the test you added for stripping the `en` subdomain from Wikipedia is a novel usecase. It seems biased towards English-speaking users to add that to the list of "common" subdomains that are stripped, so my thought is to strip a single leading subdomain, so that `wikipedia.org` would autocomplete for `de.wikipedia.org` or `ja.wikipedia.org`, for example. Alternatively we could strip _all_ subdomains but that will require extra checks around IP addresses.

**Comment by MatthewTighe** [comment]
Here's the same tests around that long reddit path for verification. They pass as is on this branch:

```
        history.recordVisit("https://www.reddit.com/r/vancouver/comments/quu9lt/hwy_1_just_north_of_lytton_is_gone", PageVisit(VisitType.LINK))
        res = history.getAutocompleteSuggestion("reddit.com/r")!!
        assertEquals("reddit.com/r/", res.text)
        assertEquals("https://www.reddit.com/r/", res.url)
        res = history.getAutocompleteSuggestion("reddit.com/r/van")!!
        assertEquals("reddit.com/r/vancouver/", res.text)
        assertEquals("https://www.reddit.com/r/vancouver/", res.url)
        res = history.getAutocompleteSuggestion("reddit.com/r/vancouver/comments/q")!!
        assertEquals("reddit.com/r/vancouver/comments/quu9lt/", res.text)
        assertEquals("https://www.reddit.com/r/vancouver/comments/quu9lt/", res.url)
        res = history.getAutocompleteSuggestion("reddit.com/r/vancouver/comments/quu9lt/h")!!
        assertEquals("reddit.com/r/vancouver/comments/quu9lt/hwy_1_just_north_of_lytton_is_gone", res.text)
        assertEquals("https://www.reddit.com/r/vancouver/comments/quu9lt/hwy_1_just_north_of_lytton_is_gone", res.url)
```

**Comment by grigoryk** [comment]
> That said, the test you added for stripping the en subdomain from Wikipedia is a novel usecase. It seems biased towards English-speaking users to add that to the list of "common" subdomains that are stripped, so my thought is to strip a single leading subdomain, so that wikipedia.org would autocomplete for de.wikipedia.org or ja.wikipedia.org, for example. Alternatively we could strip all subdomains but that will require extra checks around IP addresses.

Thinking a bit more about this, let's not go down this path. Desktop doesn't actually autocomplete in these cases, and I can see how this could be a potential security issue or just a source of weird/unexpected behaviour (e.g. silently taking a user to a page they never intended to visit, `evil.commondomain.com/some/innocuous` for input of `commondomain.com/some/inn`). Currently we do this for `m`, `www` and `mobile`, which seems innocent enough and helps us with the really common scenarios.

**Review by grigoryk — APPROVED** [review]
🚢

**Comment by grigoryk** [comment]
To close the loop, for future reference - indeed, A-S was doing path segment walking for us - so the matches we are getting back from places are already path-aware. So perhaps, the fix here wasn't strictly necessary - I filed https://github.com/mozilla/application-services/issues/4725 to look into moving it into A-S.
